Proposal add

vhe182 opened the pull request:
@meteorfox and Besan (can't link you for some reason, Besan), please provide feedback about the proposal.  

- The Rationale section needs to be written.  I think it best if one of you argues here instead of me.
- I left placeholders for implementation because we still haven't talked about how we will break down the larger tasks. Don't see much reason to have this meeting until we get the go-ahead from up-above.

vhe182 committed:
Write Abstract and Background section proposal

Also included the call for the diagram.

vhe182 committed:
Create diagram for architecture of proposal

vhe182 committed:
proposal: Add to the sections

Sections written to:
1. Proposal
1. Implementation
1. Issues

vhe182 committed:
proposal: Add to the proposal

Added to the following sections:

* Proposal/Server - Clarified statements
* Proposal/Dashboard - Initiated
* Rationale - Added topics to write about
* Open Issues - Added topics to write about

vhe182 committed:
proposal: Add text to the proposal sections

Added to the following:
- Proposal/Server API
- Proposal/Dashboard
- Issues

vhe182 committed:
proposal: Add text to proposal sections

Added to the following line:
- Implementation: added placeholder text since not sure what to put

vhe182 committed:
Merge branch 'master' of github.com:rcbops-qe/rpc-perf-dash

meteorfox commented:
I'll take a stab at it later. But we still want the implementation section.
It's high level, like, steps needed to take to start building this, which
modules will be implemented first and the what will follow.

Rationale is not just about technology, it's why this design is needed. A
good discussion here is to talk about how grafana (alternative solution)
could be used and explain pros/cons of this approach compared to the
proposal solution.

On Wed, Jan 18, 2017, 12:00 PM Victor H Estrada <notifications@github.com>
wrote:

> @meteorfox <https://github.com/meteorfox> and Besan (can't link you for
> some reason, Besan), please provide feedback about the proposal.
>
>    - The Rationale section needs to be written. I think it best if one of
>    you argues here instead of me.
>    - I left placeholders for implementation because we still haven't
>    talked about how we will break down the larger tasks. Don't see much reason
>    to have this meeting until we get the go-ahead from up-above.
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/rcbops-qe/rpc-perf-dash/pull/6
> Commit Summary
>
>    - Write Abstract and Background section proposal
>    - Create diagram for architecture of proposal
>    - Merge branch 'master' of github.com:rcbops-qe/rpc-perf-dash
>    - proposal: Add to the sections
>    - proposal: Add to the proposal
>    - proposal: Add text to the proposal sections
>    - proposal: Add text to proposal sections
>
> File Changes
>
>    - *M* docs/proposals/mvp-dashboard-design.md
>    <https://github.com/rcbops-qe/rpc-perf-dash/pull/6/files#diff-0> (74)
>
> Patch Links:
>
>    - https://github.com/rcbops-qe/rpc-perf-dash/pull/6.patch
>    - https://github.com/rcbops-qe/rpc-perf-dash/pull/6.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rcbops-qe/rpc-perf-dash/pull/6>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOLbDjm3De-i5gH-3taMWtitHH7yGFRks5rTm96gaJpZM4LnU7v>
> .
>

besanradwan commented:
Hey Victor, initially I wanted to add the comment that perhaps we should have a note about standardizing the input that will be taken in by the API. Perhaps we could use an asterisk to denote this.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,82 @@ can query the relevant test builds.
+Adding query build functionality for the user is a feature that will be
+developed in future development cycles.
+
+The server polls the data storage unit for performance test logs. The frequency
```
nitpick: drop 'unit'.

The server actually queries on-demand, it's not periodically polling. When a user loads the dashboard requests are sent to the server to load the data according to the 'filters' and environment selected by the user.

vhe182 replied:
- 'unit' dropped
- changed the text to clarify that the server is querying on demand.  My bad, I misunderstood how the server communicated with storage.  See the new commits for updated text.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,82 @@ can query the relevant test builds.
+cache of frequently and recently accessed data. The server expects a json
+result from the data store.
+
+This data can then be received by the dashboard.
```
This sentence can be safely removed. It follows from the previous discussion.

vhe182 replied:
- sentence removed

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,82 @@ can query the relevant test builds.
+The query is made via HTML requests in the form of `GET /api/labs`,
+`GET /api/labs/{lab id}`. The expected result is a json with the appropriate
+test data. This json is then used to create the graph of the specified metrics
+as they trend over the past 50 builds.
```
nitpick: Don't use '50', simply say something like 'previous builds'

vhe182 replied:
50 removed.  the sentence is now non-specific about number of builds.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,82 @@ can query the relevant test builds.
+Our choice of developing a REST API provides the benefit that any graph
+configured by a user can be sured via the URL link to that page.  This allows
+for the fast sharing of performance test results between team members and
+other teams.
```
Shareable links and the decision to use a RESTful API are not related. The implementation for 'shareable' links will probably come from the Angular application itself, not the backend API.

I don't think we need to argue why we choose a RESTful API. The obvious benefits, it's that you have a machine-consumable interface to easily build other tools and applications on top of the API, and it abstracts the storage implementation so user don't directly depend on our choice of data storage system.

vhe182 replied:
removed paragraph.

vhe182 replied:
Removed the argument for API choice.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,82 @@ can query the relevant test builds.
+    complexity and added time that the perf team will need to develop a format
+    that can accomodate all of the KPIs and their varying metrics.
+1. Should the server poll the data storage unit at a set rate or should the
+   data storage unit push updates whenever they occur?
```
Not really an issue. See comments above.

vhe182 replied:
Correct, I had forgotten during our discussion the strategy that was decided, my bad.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,82 @@ can query the relevant test builds.
+    ported to another data storage system. An issue with this strategy is that
+    the data will have to all follow the same formatting standards, this adds
+    complexity and added time that the perf team will need to develop a format
+    that can accomodate all of the KPIs and their varying metrics.
```
This sounds more like a discussion of alternative not open issues.

vhe182 replied:
Understood and removed it, thanks for pointing it out to me.

vhe182 replied:
I have moved this to rationale with appropriate reasoning.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,82 @@ can query the relevant test builds.
 
-A discussion of issues relating to this proposal for which the author does not
-know the solution. This section may be omitted if there are none.
+1. What data storage service should be used to store the test results?
```
I don't think this is an open issue, this is part of exactly what we will be solving and addressing with this proposal, this should not be left as an open issue.

vhe182 replied:
You are write, this isn't an issue.  I added rationale for the current proposed solution.  @besanradwan will add a counterargument soon as well.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,82 @@ can query the relevant test builds.
+  * [D3][2]
+  * [Angular Chart][3]
+  * [Grafana][4]
+  * Other tool?
```
Again, our proposal should not leave this as an open issue. This needs to be addressed by it.

vhe182 replied:
Removed this but also added rationale for why we are building a front end instead of using Grafana.

vhe182 committed:
proposal: Make change from review feedback

-nitpick: remove unit

Also changed the Go link at the top of the proposal so that the link is
referenced from the bottom list of references.

vhe182 committed:
proposal: Make changes as requested in feedback

Changed the body of the Proposal/Server section. The
server will not poll the data storage as previously stated.
With this commit, the text clarifies that the server will send
requests as the user makes them.

vhe182 committed:
proposal: Make changes as requested by feedback

Removed the sentence transitioning to the Dashboard section.
Not necessary because the server is not polling storage for
performance metrics data. (see previous commit for more details).

vhe182 committed:
proposal: Make changes as requested by feedback

remove specific mention of the number of builds. Generalize so that
we can later change it to whatever.

vhe182 committed:
proposal: Make changes based on feedback

- Removed paragraph describing benefits of RESTful API choice because
  it didn't make sense. There isn't a link between shareable links and
  the API choice.

vhe182 committed:
proposal: Make changes as requested in feedback

- Added Dashboard rationale
- Small changes to other areas of the doc

vhe182 committed:
proposal: Make changes as requested by feedback

- Clarified the dashboard choice rationale
- Added second argument for dashboard choice
- Added rationale topics for storage strategy choice

vhe182 committed:
proposal: Make changes as requested in feedback

- Added two rationales for data storage choice
- Removed old open issues
- Added open issue concerning the data tagging scheme

vhe182 commented:
@besanradwan and @meteorfox, any other comments?  I would like to merge this soon so we can share with RPC-QE.

This can be presented during a Wednesday demo.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+
+### Dashboard Layout
+
+bird
```
bird?

meteorfox replied:
bird is the word? :laughing:

besanradwan replied:
bird?? 😄

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+
+1. **Need for SQL-like querying with optimizations**
+
+   [InfluxDB][1] is a database service optimized for temporal data, which is a
```
s/database/time series database/

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+
+1. **Need for metrics comparison of test runs across multiple labs**
+
+   Grafana lacks the ability to directly compare multiple metrics from
```
Good point. In Grafana, time range selection is for the whole dashboard, that wouldn't work in our case if we want to compare different 'build' ranges.

meteorfox replied:
I guess in theory, one could offset the x-axis, but it's kind of hacky.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
-disadvantages of the specified approach.
+### Choice to build our own front-end
+
+1. **Avoid Grafana learning curve**
```
IMHO, I don't think this is a solid argument.

Grafana really is not that hard to pick up. The main argument I see is that Grafana being a generic dashboard meant for services and systems monitoring & visualization, not really for benchmark and performance tests comparisons or the domain concepts we want to model with our proposed system. Because of this. it hurts discoverability of available workloads results, and environments, as well as, it would be hard for other users to even know which data sets were executed together or are even relevant since it will simply look as continuous time series plot where all environments are combined (if plotted naively).

besanradwan replied:
+1

besanradwan replied:
Especially since learning curve creating your own front end is arguably steeper.

meteorfox commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
-![Diagram 1](../images/perf-dash-proposal-diagram.png)
+![alt text](../images/perf-dash-proposal-diagram.png "Perf-Dash Diagram")
+
+### Server
```
I think these sections are good but it feels like is missing a part that introduces the system as a whole and what it's set out to solve.

besanradwan replied:
Why is there a two way arrow from data store to query API?

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -1,5 +1,5 @@
 *This proposal template was taken from:*
```
nitpick: I don't think this line is needed.

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
-![Diagram 1](../images/perf-dash-proposal-diagram.png)
+![alt text](../images/perf-dash-proposal-diagram.png "Perf-Dash Diagram")
+
+### Server
```
I don't know that this is a good name for the section. Instead perhaps architecture layout/overview Additionally, I feel it is confusing to have this section and then later on subsection server layout. It feels scattered.

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+### Server
+
+The system we propose requires the development of a server API that will be
+used to serve queries that a user makes via the dashboard. To speed up the
```
I think we shouldnt be talking in first person, it should be an objective plan. Therefore it could be changed to:
"The system proposed includes many components outlined in the diagram above. A breakdown of these components....:
-API: This API serves queries that a user makes via the dashboard...."

This is just an example, but should be done throughout the document. Additionally, you will notice that I didn't say "we will be making the API". This is because I think this part is open ended and we don't know if we actually will be making the API (though that is the plan for now). Perhaps at the end of the document add a discussion session outlining the components and the possible pros and cons of different implementations. In this section I can do a pro con for InfluxAPI and you can do one for a flask or go one.

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+Adding query build functionality for the user is a feature that will be
+developed in future development cycles.
+
+When the server receives input from the user (a query), the server will send
```
nitpick: Instead of "When the server" instead say "The server receives input from..." also, instead of the "server will send" say "the server then sends..."

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+
+When the server receives input from the user (a query), the server will send
+these queries to the data storage.  The data storage returns the filtered
+performance test results. These results are then pushed to the dashboard.
```
I am a very visual person. I think the diagram you have is great! Why not add another diagram showing the data flow? I wouldnt mind making this myself if you want. In addition, if you dont want a diagram perhaps break it up into bullet points so it is easier to read.

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+When the server receives input from the user (a query), the server will send
+these queries to the data storage.  The data storage returns the filtered
+performance test results. These results are then pushed to the dashboard.
+The server expects a json result from the data store. The format and structure
```
Again these specifics, e.g. "the server expects json result" may not be the best, because we dont know that this is true at this point. I.e. we are not sure of the format yet.

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+development of the perf-dash.
+
+### Dashboard
+
```
I like how this section outlines the purpose of the dashboard and how it interacts with the other components. I will say again though that I am unsure of the specifics you are mentioning (especially the query API format if we make our own API)

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+The query is made via HTML requests in the form of `GET /api/labs`,
+`GET /api/labs/{lab id}`. The expected result is a json with the appropriate
+test data. This json is then used to create the graph of the specified metrics
+as they trend over past builds.
```
Would you mind if I did a mock up for the dashboard so we can include this as an image and reference it in this section?

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+   test lab. The team is familiar with it and we have templated various
+   visualizations for tracking the relevant metrics from each node in our
+   cloud. Using Grafana still requires time to use, time that other teams do
+   not have to invest. Our goal with the perf-dash is to allow other teams to
```
Following off of @meteorfox 's point, I dont agree that this is the rationale, but I wanted to add that we should probably make a section for choice of making our own API vs using Influx DB API

besanradwan replied:
I would want to add to this section

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
+   InfluxDB server. Uniformity between our past, current, and future datasets
+   will ensure that we can make comparisons between different test results.
+   We can accomodate new and different data by properly using tags, a feature
+   of the InfluxDB query language.
```
Great points! Perhaps incorporate how we dont have a lot of data (at least thats what we are anticipating) and thus dont need a heavy DB such as MySQL.

Additionally, we dont plan on doing massive, complex queries, and our expected data set (at least for the moment) is fairly simple therefore we dont have a need for the features that one would typically go to mySQL or SQL type DBs for. Such as secondary indexes. Also, influx DB is schema free which works to our use-case. I can cite many more things, but these are just some of the ones I think are important to include.

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
 ## Implementation
 
-A description of the steps in the implementation, who will do them, and when.
+@meteorfox has volunteered to set up the scaffolding for both the server and
```
I dont think we should have names, especially since I may want to help work on the scaffolding as well (for learning purposes)

besanradwan commented on docs/proposals/mvp-dashboard-design.md:
```diff
@@ -43,18 +43,125 @@ can query the relevant test builds.
 
-## Open issues (if applicable)
+1. Currently, the performance data we have has not been tagged to distinguish
+   the different metrics, test type, or build run. To be able to query by
```
It is with influx DB publisher for PKB. E.g. we have test type, timestamp, etc. However, the build run which presumably is linked to the release tag will be through jenkins. Maybe say this here as well.

besanradwan reviewed (changes requested):
Great work @vhe182 !!!
